03-intro-to-python.Rmd

```diff
@@ -1,9 +1,298 @@
+
+```{python}
+d = {"key1": 1,
+     "key2": 2}
+
+d["key1"] 
+#> 1
+d["key3"] = 3
+d 
+#> {'key1': 1, 'key2': 2, 'key3': 3}
+```
+
+#### Sets 
+
+Sets are used to track unique items, and can be constructed using `{val1, val2}`.
+
+```{python}
+s = {1, 2, 3}
+
+type(s)
+#> <class 'set'>
+s
+#> {1, 2, 3}
+```
+
+### Iteration with for loops
+ 
+The `for` statement in Python is similar to the `for` loop in R. It can be used to iterate over any kind of data structure.
+
+```{python}
+for x in [1, 2, 3]:
+  print(x)
+#> 1
+#> 2
+#> 3
+```
+
+### Functions
+
+Python functions are defined with the def statement. The syntax for specifying function arguments and default values is very similar to R.
+
+```{python}
+def my_function(name = "World"):
+  print("Hello", name)
+
+my_function()
+#> Hello World
+my_function("Friend")
+#> Hello Friend
+```
+
+The equivalent R code would be
+
+```{r}
+my_function <- function(name = "World") {
+  cat("Hello", name, "\n")
+}
+
+my_function()
+#> Hello World
+my_function("Friend")
+#> Hello Friend
+```
+
+
+### Classes and Object Oriented Programming (OOP)
+
+In R, the most widely used unit of composition for code is functions, and in Python, it is classes. Classes are how you organize and find methods in Python. This approach to code composition is called object oriented programming (OOP). Let's dive in the details of OOP.
+
+An object is any entity that you want to store and process data about. Each object is an instance of a class in the computer's memory. A class is a template for creating objects. Creating an object from a class is called instantiation. It has properties and methods (functions for the class). 
+
+For example, we could have a class called Person. The properties of this class are what describe this Person class:
+
+- first_name
+- last_name
+- gender
+- date_of_birth
+- occupatiaon
+
+
+The methods of this class are the functions for this Person class:
+
+- walk()
+- run()
+- sleep()
+- eat()
+
+
+Here is a simple Person class for demonstration purposes.
+
+```{python}
+class Person:
+  pass # `pass` means do nothing.
+
+Person
+#> <class '__main__.Person'>
+type(Person)
+#> <class 'type'>
+
+instance = Person()
+instance
+#> <__main__.Person object at 0x102ba75e0>
+type(instance)
+#> <class '__main__.Person'>
+```
+
+Like the `def` statement, the `class` statement is used to create a Python class. First note the strong naming convention, classes are typically CamelCase, and functions are typically snake_case. After defining Person, you can interact with it, and see that it has type 'type'. Calling `instance = Person()` creates a new object instance of the class, which has type `Person` (ignore the __main__. prefix for now). 
+
+
+### Importing modules
+
+In R, authors can bundle their code into R packages, and R users can access objects from R packages via `library()` or `::`. In Python, authors bundle code into modules, and users access modules using `import`.
```

Comment:
Nit: pip-installable libraries are also called "packages" in Python. Modules are the individual `.py` files inside a package.

Reply:
Good point. Thanks for the note.